File: algorithms/2_medium/R/00061_rotate_list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    static inline pair<int, int> calc_len_and_shift_count(ListNode* head, int k) {
        int len{ 0 };
        auto curr{ head };

        while (curr != nullptr) {
            len += 1;
            curr = curr->next;
        }

        const int shift_count{ len == 0 ? 0 : (len - k % len) % len };

        return {len, shift_count};
    }

    static inline pair<ListNode*, ListNode*> split_into_two_list(ListNode* head, int shift_count) {
        auto curr{ head };

        for (int i{ 0 }; i < shift_count - 1; ++i) {
            curr = curr->next;
        }

        auto head2{ curr->next };
        curr->next = nullptr;

        return {head, head2};
    }
public:
    static ListNode* rotateRight(ListNode* head, int k) {
        const auto [len, shift_count] = calc_len_and_shift_count(head, k);

        if (len <= 1 || shift_count == 0) {
            return head;
        }

        auto [h1, h2] = split_into_two_list(head, shift_count);
        auto curr{ h2 };

        while (curr->next != nullptr) {
            curr = curr->next;
        }

        curr->next = h1;

        return h2;
    }
};

File: algorithms/2_medium/R/00061_rotate_list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def calc_len_and_shift_count(self, head: Optional[ListNode], k: int) -> (int, int):
        n = 0
        curr = head

        while curr != None:
            n += 1
            curr = curr.next

        shift_count = 0 if n == 0 else (n - k % n) % n

        return (n, shift_count)

    def split_into_two_list(
        self,
        head: Optional[ListNode],
        shift_count: int
    ) -> (Optional[ListNode], Optional[ListNode]):
        curr = head

        for i in range(shift_count - 1):
            curr = curr.next

        head2 = curr.next
        curr.next = None

        return (head, head2)

    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        (n, shift_count) = self.calc_len_and_shift_count(head, k)

        if n <= 1 or shift_count == 0:
            return head

        (h1, h2) = self.split_into_two_list(head, shift_count)
        curr = h2

        while curr.next != None:
            curr = curr.next

        curr.next = h1

        return h2
